### PR TITLE
Setup OAuth proxy for e2e test

### PR DESCRIPTION
**Pull request by IniZio**
ref #3773

dependencies:
- [x] #3930

### CODE REVIEW

**Comment by IniZio** [comment]
Will add redirect step to allow testing more oaouth cases in future e.g. oauth error.

Example config:
```yaml
- action: oauth_redirect
  to: data.result.url
  redirect_uri: http://mock
  output:
    result: |
      {
        "code": "[[string]]"
      }

- action: input 
  input: |
    {
      "code": {{ prevstep.output.result.code }}
    }
```

**Comment by louischan-oursky** [comment]
You can rebase once first.

**Comment by IniZio** [comment]
Updated to run oauth redirect

File: pkg/util/httputil/client.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"net/http"
+	"net/url"
+
+	"os"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
```
Comment (louischan-oursky) [inline pkg/util/httputil/client.go:11]: util cannot import lib. Create a new struct in this package instead.

File: pkg/util/httputil/client.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+)
+
+func NewHTTPClient(env *config.EnvironmentConfig) *http.Client {
+	client := http.DefaultClient
```
Comment (louischan-oursky) [inline pkg/util/httputil/client.go:15]: Do not modify DefaultClient. According to https://pkg.go.dev/net/http#Client a zero value of Client is good enough.

File: pkg/lib/authn/sso/access_token.go [hunk 3]
```diff
@@ -78,6 +78,7 @@ func (r AccessTokenResp) TokenType() string {
 }
 
 func fetchAccessTokenResp(
+	client *http.Client,
```
Comment (louischan-oursky) [inline pkg/lib/authn/sso/access_token.go:81]: Make a separate commit for switching to non DefaultClient.

File: pkg/lib/config/environment.go [hunk 4]
```diff
@@ -50,6 +50,10 @@ type EnvironmentConfig struct {
 	DevMode DevMode `envconfig:"DEV_MODE" default:"false"`
 	// LogLevel sets the global log level
 	LogLevel string `envconfig:"LOG_LEVEL" default:"warn"`
+	// HTTPProxy sets the HTTP proxy
+	HTTPProxy string `envconfig:"HTTP_PROXY"`
+	// SSLCertFile sets additional SSL certificate file
+	SSLCertFile string `envconfig:"SSL_CERT_FILE"`
```
Comment (louischan-oursky) [inline pkg/lib/config/environment.go:56]: The name of these variables is too general, making people think Authgear will support them. But in fact, they are used in e2e test only. I suggest we make the authgear command to take 2 extra CLI options, `--e2e-http-proxy-url` and `--e2e-tls-ca-cert-path`. When the CLI options are given, then the HTTP client used by OAuth will be configured accordingly.

File: pkg/util/httputil/deps.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ import "github.com/google/wire"
 
 var DependencySet = wire.NewSet(
 	NewJSONResponseWriterLogger,
+	NewHTTPClient,
```
Comment (louischan-oursky) [inline pkg/util/httputil/deps.go:7]: Wire inject by type. Adding NewHTTPClient will make it available in the graph, other components can then use it. But it is not our intention. So add a new type `type OAuthHTTPClient *http.Client` in `pkg/lib/authn/sso`. Add `NewOAuthHTTPClient` in `pkg/lib/authn/sso/deps.go`. Make OAuthProviderFactory to use OAuthHTTPClient instead of *http.Client.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/util/httputil/client.go (hunk 1 above)
Comment (IniZio) [reply pkg/util/httputil/client.go]: Moved http client inside `authn/sso`

File: pkg/util/httputil/deps.go (hunk 5 above)
Comment (IniZio) [reply pkg/util/httputil/deps.go]: Think this will make `OAuthHTTPClient` unable to use methods of `*http.Client` 

If I use anonymous field though, I won't be able to intercept it with `gock.InterceptClient`.

File: e2e/pkg/testrunner/testcase.go [hunk 6]
```diff
@@ -249,13 +249,19 @@ func execTemplate(cmd *End2EndCmd, prev *StepResult, content string) (string, er
 	tmpl := texttemplate.New("")
 	tmpl.Funcs(makeTemplateFuncMap(cmd))
 
-	_, err := tmpl.Parse(content)
+	// Treat ${{ as {{ for oneline templates
+	_, err := tmpl.Parse(strings.ReplaceAll(content, "${{", "{{"))
```
Comment (louischan-oursky) [inline e2e/pkg/testrunner/testcase.go:253]: Do not add this.

Instead, we can write

```yaml
to: |-
  {{ .some.value }}
```

File: pkg/util/httputil/deps.go (hunk 5 above)
Comment (IniZio) [reply pkg/util/httputil/deps.go]: Updated to use custom struct for sso, thanks!

**Review by louischan-oursky — APPROVED** [review]
